<a href="https://colab.research.google.com/github/wiv33/A-Learning-python/blob/master/machine-learning/_000_hello_machine/_000_basic/_003_tensorflow_cuk_edu/_002_2_ml/_002_weather_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 기상청 날씨 데이터 분석하기

## 데이터 불러오기

- 컬럼 설정
- 사람을 위한 월 시작 row 제거
  - 12월 31일 -> | start: 201912 | -> 1월 1일

In [118]:
import pandas as pd
path = 'https://raw.githubusercontent.com/wiv33/A-Learning-python/master/machine-learning/_000_hello_machine/_000_basic/_003_tensorflow_cuk_edu/_002_2_ml/data'

df_precipitation = pd.read_csv(path + '/precipitation_201910_202009.csv')
df_precipitation.columns = ['day', 'hour', 'precipitation']
df_precipitation.dropna(axis=0, inplace=True)

df_direction_of_the_wind = pd.read_csv(path + '/direction_of_the_wind_201910_202009.csv')
df_direction_of_the_wind.columns = ['day', 'hour', 'wind']
df_direction_of_the_wind.dropna(axis=0, inplace=True)

df_precipitation_shape = pd.read_csv(path + '/precipitation_shape_201910_202009.csv')
df_precipitation_shape.columns = ['day', 'hour', 'precipitation_shape']
df_precipitation_shape.dropna(axis=0, inplace=True)

df_temperature = pd.read_csv(path + '/temperature_201910_202009.csv')
df_temperature.columns = ['day', 'hour', 'temperature'] 
df_temperature.dropna(axis=0, inplace=True)

df_humidity = pd.read_csv(path + '/humidity_201910_202009.csv')
df_humidity.columns = ['day', 'hour', 'humidity']
df_humidity.dropna(axis=0, inplace=True)

df_wind_spped = pd.read_csv(path + '/%EC%9B%94%EA%B3%A1%EC%A0%9C1%EB%8F%99_%ED%92%8D%EC%86%8D_wind_speed_201910_202009.csv')
df_wind_spped.columns = ['day', 'hour', 'wind_speed']
df_wind_spped.dropna(axis=0, inplace=True)


nan row 확인

In [119]:
df_test = pd.read_csv(path + '/temperature_201910_202009.csv')
# df_test.iloc[3318:3340, :]

In [120]:
df_test2 = pd.read_csv(path + '/precipitation_shape_201910_202009.csv')
# df_test2.iloc[3318:3340, :]

In [121]:
df_test3 = pd.read_csv(path + '/precipitation_201910_202009.csv')
# df_test3.iloc[3318:3340, :]

In [122]:
df_test4 = pd.read_csv(path + '/precipitation_201910_202009.csv')
df_test4.columns = ['day', 'hour', 'precipitation']
df_test4.head()



day   hour  precipitation
0   1    0.0            0.0
1   1  100.0            0.0
2   1  200.0            0.0
3   1  300.0            0.0
4   1  400.0            0.0

In [123]:
df_date_test4 = pd.date_range('2019-10-01', periods=366)
df_date_test4

DatetimeIndex(['2019-10-01', '2019-10-02', '2019-10-03', '2019-10-04',
               '2019-10-05', '2019-10-06', '2019-10-07', '2019-10-08',
               '2019-10-09', '2019-10-10',
               ...
               '2020-09-21', '2020-09-22', '2020-09-23', '2020-09-24',
               '2020-09-25', '2020-09-26', '2020-09-27', '2020-09-28',
               '2020-09-29', '2020-09-30'],
              dtype='datetime64[ns]', length=366, freq='D')

nan 형태 확인 후 제거

In [124]:
del df_test
del df_test2
del df_test3
del df_date_test4

### 다섯 개의 DataFrame shape 확인하기

In [125]:
df_direction_of_the_wind.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   day     8784 non-null   object 
 1   hour    8784 non-null   float64
 2   wind    8784 non-null   float64
dtypes: float64(2), object(1)
memory usage: 274.5+ KB


In [126]:
df_precipitation_shape.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 3 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   day                  8784 non-null   object 
 1   hour                 8784 non-null   float64
 2   precipitation_shape  8784 non-null   float64
dtypes: float64(2), object(1)
memory usage: 274.5+ KB


In [127]:
df_humidity.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   day       8784 non-null   object 
 1   hour      8784 non-null   float64
 2   humidity  8784 non-null   float64
dtypes: float64(2), object(1)
memory usage: 274.5+ KB


In [128]:
df_temperature.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   day          8784 non-null   object 
 1   hour         8784 non-null   float64
 2   temperature  8784 non-null   float64
dtypes: float64(2), object(1)
memory usage: 274.5+ KB


8784 row와 3개의 컬럼 중 세 번째 컬럼을 하나로 만들면 될 것.

## 각 데이터 프레임 하나로 합치기

강수, 강수형태, 습도, 기온, 풍향, 풍속

In [129]:
df = df_precipitation.iloc[:, :2]
df['precipitation_강수'] = df_precipitation.precipitation
df['precipitation_shape_강수형태'] = df_precipitation_shape['precipitation_shape']
df['temperature_기온'] = df_temperature['temperature']
df['humidity_습도'] = df_humidity.humidity
df['wind_풍향'] = df_direction_of_the_wind.wind
df['wind_spped_풍속'] = df_wind_spped.wind_speed


### merge 결과 확인하기

In [130]:
df.head()

day   hour  precipitation_강수  ...  humidity_습도  wind_풍향  wind_spped_풍속
0   1    0.0               0.0  ...         63.0     29.0            0.8
1   1  100.0               0.0  ...         60.0    183.0            1.6
2   1  200.0               0.0  ...         56.0    122.0            1.6
3   1  300.0               0.0  ...         57.0     64.0            1.9
4   1  400.0               0.0  ...         59.0    127.0            1.5

[5 rows x 8 columns]

## 시간 데이터 타입 변환

- 기존 24시의 표현 변경

  0.0, 100.0, 200.0

  0, 1, 2, 3

- ~~day int로 변경~~
- day를 datetime 형식으로 변경.

시각화할 때 구분이 가지 않음.


In [131]:
def hour_regulation(hour: float):
  return int(hour / 100)

In [132]:
df_day_idx = pd.date_range('2019-10-01', periods=366)

In [133]:
df_day_idx

DatetimeIndex(['2019-10-01', '2019-10-02', '2019-10-03', '2019-10-04',
               '2019-10-05', '2019-10-06', '2019-10-07', '2019-10-08',
               '2019-10-09', '2019-10-10',
               ...
               '2020-09-21', '2020-09-22', '2020-09-23', '2020-09-24',
               '2020-09-25', '2020-09-26', '2020-09-27', '2020-09-28',
               '2020-09-29', '2020-09-30'],
              dtype='datetime64[ns]', length=366, freq='D')

In [148]:

def day_regulation(day):
  
  return temp

In [150]:
df.hour = df.hour.apply(hour_regulation)
# df.index = df_day_idx
# df.day = df.day.apply(day_regulation)

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   day                       8784 non-null   object 
 1   hour                      8784 non-null   int64  
 2   precipitation_강수          8784 non-null   float64
 3   precipitation_shape_강수형태  8784 non-null   float64
 4   temperature_기온            8784 non-null   float64
 5   humidity_습도               8784 non-null   float64
 6   wind_풍향                   8784 non-null   float64
 7   wind_spped_풍속             8784 non-null   float64
dtypes: float64(6), int64(1), object(1)
memory usage: 617.6+ KB


In [ ]:
df.describe()

### 이상 데이터 확인

강수, 강수 형태, 습도, 풍향, 풍속의 최소 값이 `-1`이고

기온이 `-50도`인 것은 정규화가 필요하다고 판단.

In [ ]:
df[df['temperature_기온'] == -50].index

- 이상 데이터의 인덱스 확인

### 이상 데이터 정규화

- 습도와 기온은 +-2일 값의 평균으로 적용
- 나머지 값은 0으로 적용

In [ ]:
import numpy as np 

for x in df[df['wind_풍향'] <= -1].index:
  seq = int(x)
  df['temperature_기온'][seq] = np.mean((df.iloc[seq-2:seq, 4], df.iloc[seq+1:seq+3, 4]))
  df['humidity_습도'][seq] = np.mean((df.iloc[seq-2:seq, 5], df.iloc[seq+1:seq+3, 5]))
  df['wind_spped_풍속'][seq] = 0.0
  df['precipitation_shape_강수형태'][seq] = 0.0
  df['precipitation_강수'][seq] = 0.0
  df['wind_풍향'][seq] = 0.0


In [ ]:
df[df['humidity_습도'] <= 10.0]

In [ ]:
df[df['temperature_기온'] <= -50]

In [ ]:
df.iloc[1055: 1057, :]

In [ ]:
df[df['wind_풍향'] <= -1]

In [ ]:
df.describe()

In [ ]:
df[df['precipitation_shape_강수형태'] == 3.]

# 시각화

In [ ]:
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

## 시간에 따른 기온의 변화

In [ ]:
df.head()

In [ ]:
df.tail()

In [ ]:
plt.figure(figsize=(12, 5))


plt.title('Temperature variation over time')
plt.xlabel('hour')
plt.ylabel('temperature')

sns.barplot(x=df['day'], y=df['temperature_기온'])

In [ ]:
sns.distplot(df['temperature_기온'],
             hist=True,
             kde=True,
             rug=True)